### PR TITLE
Move On*Update handling into sync funcion

### PR DESCRIPTION
**Pull request by thockin**

### CODE REVIEW

**Comment by wojtek-t** [comment]
@thockin - can you please those 4 commits into a separate PR:
- Rename functions for congruence
- Return healthcheck info from buildNewEndpointsMap
- save allServices in prep for async iptables
- add a reason code to syncProxyRules
Those are useful anyway in my opinion, and we can merge those now.

**Comment by thockin** [comment]
I broke out PRs #44056 #44051 #44053 #44055 

Some of them stack, so it's not quite perfect, but you can see them in isolation.

File: pkg/proxy/iptables/proxier.go [hunk 1]
```diff
@@ -753,6 +722,23 @@ func (proxier *Proxier) syncProxyRules(reason syncReason) {
 		glog.V(2).Info("Not syncing iptables until Services and Endpoints have been received from master")
 		return
 	}
+
+	// Figure out the new services we need to activate.
+	newServices, hcServices, staleServices := buildNewServiceMap(proxier.allServices, proxier.serviceMap)
+
+	// If this was called because of a services update, but nothing actionable has changed, skip it.
+	if reason == syncReasonServices && reflect.DeepEqual(newServices, proxier.serviceMap) {
+		glog.V(3).Infof("Skipping iptables sync because nothing changed")
```
Comment (wojtek-t) [inline pkg/proxy/iptables/proxier.go]: return ?

File: pkg/proxy/iptables/proxier.go [hunk 2]
```diff
@@ -753,6 +722,23 @@ func (proxier *Proxier) syncProxyRules(reason syncReason) {
...
+	}
+
+	// Figure out the new endpoints we need to activate.
+	newEndpoints, hcEndpoints, staleEndpoints := buildNewEndpointsMap(proxier.allEndpoints, proxier.endpointsMap, proxier.hostname)
+
+	// If this was called because of an endpoints update, but nothing actionable has changed, skip it.
+	if reason == syncReasonEndpoints && reflect.DeepEqual(newEndpoints, proxier.endpointsMap) {
+		glog.V(3).Infof("Skipping iptables sync because nothing changed")
```
Comment (wojtek-t) [inline pkg/proxy/iptables/proxier.go]: return ?

**Comment by wojtek-t** [comment]
So it seems that only last 2 commits are not yet merged. @thockin do you think we can proceed with this PR now?

File: pkg/proxy/iptables/proxier.go (hunk 1 above)
Comment (thockin) [inline pkg/proxy/iptables/proxier.go]: Oops!  That is why we do code review, I guess.

**Comment by thockin** [comment]
rebased and re-pushed

**Comment by wojtek-t** [comment]
/lgtm

@thockin - can we remove "do-not-merge" label?

**Comment by thockin** [comment]
This can go in. I started merging the async rate-limiting logic, but it's somewhat more complicated now that we do no-op detection for Update calls.  Have to rethink how to do it best.

**Comment by thockin** [comment]
@k8s-bot gci gce e2e test this
@k8s-bot gce etcd3 e2e test this

**Comment by wojtek-t** [comment]
@k8s-bot gci gce e2e test this

**Pull request merged**